Corona-Warn-App/src/main/java/de/rki/coronawarnapp/diagnosiskeys/download/KeyFileDownloader.kt

```diff
@@ -0,0 +1,303 @@
+
+        val storageResult = deviceStorage.checkSpacePrivateStorage(
+            // 512KB per day file, for 15 days, for each country ~ 65MB for 9 countries
+            requiredBytes = intersection.size * 15 * 512 * 1024L
+        )
+
+        Timber.tag(TAG).d("Storage check result: %s", storageResult)
+        if (!storageResult.isSpaceAvailable) {
+            throw IOException("Not enough free space (${storageResult.freeBytes}")
+        }
+
+        val availableKeys = if (CWADebug.isDebugBuildOrMode && settings.isLast3HourModeEnabled) {
+            syncMissing3Hours(currentDate, intersection)
+            keyCache.getEntriesForType(CachedKeyInfo.Type.COUNTRY_HOUR)
+        } else {
+            syncMissingDays(intersection)
+            keyCache.getEntriesForType(CachedKeyInfo.Type.COUNTRY_DAY)
+        }
+
+        return@withContext availableKeys
+            .filter { it.first.isDownloadComplete && it.second.exists() }
+            .mapNotNull { (keyInfo, path) ->
+                if (!path.exists()) {
+                    Timber.tag(TAG).w("Missing keyfile for : %s", keyInfo)
+                    null
+                } else {
+                    path
+                }
+            }
+            .also { Timber.tag(TAG).d("Returning %d available keyfiles", it.size) }
+    }
+
+    /**
+     * Fetches files given by serverDates by respecting countries
+     * @param availableCountries pair of dates per country code
+     */
+    @Suppress("LongMethod")
+    private suspend fun syncMissingDays(
+        availableCountries: List<LocationCode>
+    ) = withContext(Dispatchers.IO) {
+        val availableCountriesWithDays = availableCountries.map {
+            val days = diagnosisKeyServer.getDayIndex(it)
+            CountryDays(it, days)
+        }
+
+        val cachedDays = keyCache
+            .getEntriesForType(CachedKeyInfo.Type.COUNTRY_DAY)
+            .filter { it.first.isDownloadComplete && it.second.exists() } // We overwrite not completed ones
+            .map { it.first }
+
+        // All cached files that are no longer on the server are considered stale
+        val staleKeyFiles = cachedDays.filter { cachedKeyFile ->
+            val availableCountry = availableCountriesWithDays.singleOrNull {
+                it.country == cachedKeyFile.location
+            }
+            if (availableCountry == null) {
+                Timber.tag(TAG).w(
+                    "Unknown location %s, assuming stale cache.",
+                    cachedKeyFile.location
+                )
+                return@filter true // It's stale
+            }
+
+            availableCountry.dayData.none { date ->
+                cachedKeyFile.day == date
+            }
+        }
+        if (staleKeyFiles.isNotEmpty()) keyCache.delete(staleKeyFiles)
+
+        val nonStaleDays = cachedDays.minus(staleKeyFiles)
+        val countriesWithMissingDays = availableCountriesWithDays.mapNotNull {
+            it.toMissingDays(nonStaleDays)
```

Comment:
I disagree here, there is little benefit to encoding the explicit return type in the function name "Days" vs "CountryDays" and the downside that if we ever refactor `CountryDays` we have to rename the function too. Due to switching from "String" to "CountryDays" we already prevent accidentally using a different type for something.

Where do you see an exception? `mapNotNull` will filter out all elements that are null, it a combo of `map{}.filterNotnull()`.

`toMissingDays` returns missing days if there are some, or null if there are no missing days. I agree that it may be slightly unexpected to have it return `null` but I don't see the issue in handling, Kotlin will throw an error if that null is not handled correctly.